src/adjust.jl

```diff
@@ -144,3 +144,33 @@ function _adjust(r::T, nt::NamedTuple) where T <: AbstractRule
   end
   T(vals...)  # relies on having the default constructor
 end
+
+###
+#adjust with type control
+###
+
+adjust!(ℓ::Leaf, oT::Type, eta::Real) = (ℓ.rule = adjust(ℓ.rule, oT, eta); nothing)
+adjust!(ℓ::Leaf, oT::Type; kw...) = (ℓ.rule = adjust(ℓ.rule, oT; kw...); nothing)
+
+adjust(ℓ::Leaf, oT::Type, eta::Real) = Leaf(adjust(ℓ.rule, oT, eta), ℓ.state, ℓ.frozen)
+adjust(ℓ::Leaf, oT::Type; kw...) = Leaf(adjust(ℓ.rule, oT; kw...), ℓ.state, ℓ.frozen)
+
+adjust!(tree, oT::Type, eta::Real) = foreach(st -> adjust!(st, oT, eta), tree)
+adjust!(tree, oT::Type; kw...) = foreach(st -> adjust!(st, oT; kw...), tree)
+
+adjust(r::AbstractRule, oT::Type, eta::Real) = ifelse(isa(r, oT), adjust(r, eta), r)
```

Comment:
These overloads are allocating the output of `adjust(r, ...)` even if it doesn't end up being used. Unless `ifelse` brings a demonstrable type stability benefit over a plain conditional, I'd recommend using a ternary instead:
```suggestion
adjust(r::AbstractRule, oT::Type{<:AbstractRule}, eta::Real) = isa(r, oT) ? adjust(r, eta) : r
```

Reply:
I would guess the branch is resolved when compiling?

I think you could also make dispatch do this, although whether it's a good idea I don't know:
```suggestion
adjust(r::C, oT::Type{A}, eta::Real) where {C <: A <: AbstractRule}  where {C<:AbstractRule} = adjust(r, eta)
adjust(r:: AbstractRule, oT::Type{<:AbstractRule}, eta::Real) = r
```